lib/ab_test.rb

```diff
@@ -1,21 +1,31 @@
 # frozen_string_literal: true
 
-class AbTestBucket
+class AbTest
   attr_reader :buckets, :experiment_name, :default_bucket
 
   MAX_SHA = (16 ** 64) - 1
 
-  def initialize(experiment_name:, buckets: {}, default_bucket: :default)
+  def initialize(
+    experiment_name:,
+    discriminator:,
+    buckets: {},
+    default_bucket: :default
+  )
```

Comment:
I think it might be easier if we passed in `discriminator` as a block to the constructor? maybe document its interface with `@yieldparam` too

then definition of the tests at the callsites look like
```ruby
AbTest.new(
  name: 'foobar',
) do |request:, service_provider:, user:, user_session:|
  idv_thing_session[session]
end
```

Reply:
I like this!